Hotfix/implement daemonized scan

selankon opened the pull request:
* Implement daemonized scan on a subshell
* Implement `start` `stop` `status` and `restart` ubus methods
* Implement testing for those. 

After merged will be added to https://github.com/libremesh/lime-packages/pull/925 PR

Changes on lime-app will be needed.

selankon committed:
firstbootwizard: delete fbw init.d service to use `execute_daemonized` instead

selankon committed:
firstbootwizard: delete deprecated get status code

selankon committed:
firstbootwizard: refactor function name

selankon committed:
firstbootwizard: implement start, stop, status, restart functions

selankon committed:
firstbootwizard: Implement start, stop, status, restart tests

selankon committed:
firstbootwizard: Implement fbw status

selankon committed:
firstbootwizard: Implement missing local methods

germanferrero commented on packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua:
```diff
@@ -454,8 +436,57 @@ function fbw.get_all_networks()
+function fbw.status_search_networks()
+    local scan_file = fbw.check_scan_file()
+    local status
+    if (scan_file == nil) or (scan_file == "true") then
```
This looks incorrect to me. If scan_file is nil then status is not "scanning" but something like "idle", isn't it?

selankon replied:
In fact, this was implemented previously and it starts the scan. Actually `scan_file == nil`  mean scan never started, never stopped. `idle` seems a good status message for me. See 232983cc

selankon replied:
Probably this need an implementation frontend side

germanferrero commented on packages/first-boot-wizard/tests/test_firstbootwizard.lua:
```diff
@@ -178,17 +178,46 @@ describe('FirstBootWizard tests #fbw', function()
+    it('test start search networks', function()
+        fbw.end_scan()
+        local results = fbw.start_search_networks()
+        assert(true, results)
```
There are two issues here:
- The name of the test "test start search networks", should say what is actually being tested. Example: an script is executed, another function is called, it returns some expected value, etc.
- We are not testing anything with: `assert(true, results)` 
From [here](https://www.lua.org/pil/8.3.html) 
> The assert function checks whether its first argument is not false and simply returns that argument; if the argument is false (that is, false or nil), assert raises an error. Its second argument, the message, is optional, so that if you do not want to say anything in the error message, you do not have to.

You may want to use assert.is_true, assert.are_same, and such from https://olivinelabs.com/busted/#asserts

germanferrero replied:
This comments apply for the other tests aswell

selankon replied:
I rewrite all tests again reformating the code and it show me some errors on tests. Fixed after 4f7394ef

germanferrero commented on packages/first-boot-wizard/files/usr/libexec/rpcd/lime-fbw:
```diff
@@ -64,11 +79,16 @@ end
     elseif  arg[2] == 'set_network'     then set_network(msg)
     elseif  arg[2] == 'create_network'  then create_network(msg)
     elseif  arg[2] == 'dismiss'         then dismiss(msg)
+    elseif  arg[2] == 'fbw_status'     	then fbw_status(msg)
```
We could have only one status endpoint which states everything related to fbw status:
- lock: (configured with fbw or dismissed)
- status: idle, scanning, scanned.
- scanned networks
- downloaded configurations.

As first sight is difficult to me to guess what should I expect from status vs fbw_status.

selankon replied:
Indeed, this will need modifications frontend side.

selankon replied:
`fbw_status` function deprecated here 4cd06df8

selankon committed:
firstbootwizard: Fix rpcd bad syntaxis

selankon committed:
firstbootwizard: Fix tests

selankon committed:
firstbootwizard: fix bad assertion

selankon committed:
firstbootwizard: implement kill_fbw

selankon committed:
firstbootwizard: implement idle status message

selankon committed:
firstbootwizard: deprecate fbw_status function

germanferrero reviewed (approved):
Good work @selankon the code is cleaner now!

selankon committed:
firstbootwizard: fix ubus error

The error `Command failed: Invalid argument` was caused by a bad json output

selankon committed:
firstbootwizard: return correct status object

selankon committed:
firstbootwizard: use nixio to log

This fix a bug on the firstbootwizard ubus interface, that return "invalid argument" error because it printed
a debug line.

germanferrero reviewed (approved):
I'v tried it out! It works :)
